Add specification (#56)
ballerina-platform/module-ballerinax-sap.jco: docs/spec/spec.md

# Specification: Ballerina SAP JCo Connector

_Authors_: @RDPerera\
_Reviewers_: @aashikam @shafreenAnfar\
_Created_: 2024/08/28\
_Updated_: 2024/09/02\
_Edition_: Swan Lake

> ## Introduction

Reviewer: Shall we use the standard template [1] for the spec introduction section ?

[1] - https://github.com/ballerina-platform/module-ballerina-auth/blob/master/docs/spec/spec.md#introduction

Author: Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

The conforming implementation of the specification is included in the Ballerina distribution. Any deviations from the specification are considered bugs.

## Contents

1. [Introduction](#1-introduction)
2. [Components](#2-components)
   - 2.1 [RFC Client](#21-rfc-client)
     - 2.1.1 [Configurations](#211-configurations)
     - 2.1.2 [Available Operations](#212-available-operations)
       - 2.1.2.1 [Execute a Function Module via RFC](#2121-execute-a-function-module-via-rfc)
       - 2.1.2.2 [Send IDocs to an SAP System](#2122-send-idocs-to-an-sap-system)
     - 2.1.3 [Data Types](#213-data-types)
   - 2.2 [Listener](#22-listener)
     - 2.2.1 [Configurations](#221-configurations)
     - 2.2.2 [Available Operations](#222-available-operations)
       - 2.2.2.1 [Receive IDocs](#2221-receive-idocs)
       - 2.2.2.2 [Error Handling in Listener](#2222-error-handling-in-listener)

> ## 1. Introduction

Reviewer: IMO this should be `Overview`

Author: Yes. This was a mistake, Addressed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

> ## 1. Introduction

Reviewer 2: I think `Introduction` section is repeated here. Maybe we can use an `Overview` section to provide a summary of the connector spec

Author: Yes. It was a mistake. Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

The Ballerina SAP JCo Connector provides seamless integration with SAP systems through Java Connector (JCo) capabilities. This connector allows Ballerina applications to interact with SAP systems, enabling operations like Remote Function Call (RFC) execution, IDoc processing, and more. By leveraging this connector, developers can easily integrate SAP functionalities into their Ballerina applications, making it an essential tool for enterprises working with SAP.

The `ballerinax/sap.jco` package exposes the SAP JCo library as Ballerina functions, enabling easy access to SAP's enterprise resource planning (ERP) software and other SAP solutions, such as human capital management (HCM), customer relationship management (CRM), and supply chain management (SCM).

### Architecture Overview

> The SAP JCo Connector architecture consists of key components that facilitate communication between Ballerina applications and SAP systems. These components include the SAP function modules, the SAP system itself, Ballerina code that handles the RFC calls, and the IDoc servers responsible for sending and receiving IDocs. This architecture enables robust integration, supporting various enterprise operations.

Reviewer 3: ```suggestion
The SAP JCo Connector architecture comprises key components facilitating communication between Ballerina applications and SAP systems. These components include the SAP function modules, the SAP system itself, the Ballerina code that handles the RFC calls, and the IDoc servers responsible for sending and receiving IDocs. This architecture enables robust integration, supporting various enterprise operations.
```

Author: Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

## 2. Components

### 2.1 RFC Client

#### 2.1.1 Configurations

> ##### 2.1.1.1 DestinationConfig
> 
> The `DestinationConfig` type holds the configuration details needed to create a BAPI connection.

Reviewer: Instead we could do this. Check for other places as well.
```suggestion
- The `DestinationConfig` represents the configuration details needed to create a BAPI connection.
```

Author: Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

- **ashost**: The SAP host name (`jco.client.ashost`).
- **sysnr**: The SAP system number (`jco.client.sysnr`).
- **jcoClient**: The SAP client (`jco.client.client`).
- **user**: The SAP user name (`jco.client.user`).
- **passwd**: The SAP password (`jco.client.passwd`).
- **lang**: The SAP language (`jco.client.lang`). Default is "EN".

> - **group**: The SAP group (`jco.client.group`). Default is "PUBLIC".

Reviewer: Rather doing this shall we add the relevant Ballerina record with proper documentation ?

Author: Yes. That would be better, Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

**Example Configuration:**

```ballerina
configurable jco:DestinationConfig config = ?;
jco:Client sapClient = check new (config);
```

Config.toml file:

```toml
[config]
ashost = "host"
sysnr = "00"
jcoClient = "000"
user = "user"
passwd = "password"
lang = "en"
group = "group"
```

##### 2.1.1.2 AdvancedConfig

The `AdvancedConfig` type is a map that holds any custom configurations needed to create a SAP connection.

**Example Configuration:**

```ballerina
configurable jco:AdvancedConfig config = ?;
jco:Client sapClient = check new (config);

> ```

Author: Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

Config.toml file:

```toml
[config]
"jco.destination.pool_capacity" = "3"
"jco.destination.peak_limit" = "10"
```

#### 2.1.2 Available Operations

> ##### 2.1.2.1 Execute a Function Module via RFC

Reviewer: IMO the spec is for a library developer who wants to write a library package. So, here we should define the API rather than providing examples. Shall we define the API or the functions supported by this client with the request/response payload types ?

Author: Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

You can execute a specific function module on the SAP system using the `execute` function. This operation requires you to specify the function module name and the necessary parameters.

```ballerina
public function main() returns error? {
    ImportParams importParams = {
        importParam1: "Hello",
        importParam2: 123,
        importParam3: 123.45,
        importParam4: 123.456
    };

    ExportParams? result = check jcoClient->execute("TEST_FUNCTION", importParams);

>     if (result is jco:ExportParams) {

Reviewer 2: ```suggestion
    if result is jco:ExportParams {
```

Author: Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

        io:println("Result: ", result);
    } else {
        io:println("Error: Function execution failed");
    }
}
```

##### 2.1.2.2 Send IDocs to an SAP System

The RFC Client also supports sending IDocs to an SAP system, allowing you to automate the exchange of structured data.

```ballerina
public function main() returns error? {
    xml iDoc = check io:fileReadXml("resources/sample.xml");
    check jcoClient->sendIDoc(iDoc);
    io:println("IDoc sent successfully.");
}
```

#### 2.1.3 Data Type Mapping Between Ballerina and SAP JCo

> ##### 2.1.3.6 Ballerina to SAP JCo Mapping

Reviewer 3: Do we need another sub title here

Author: Yes, that is unnecessary. Changed with [1a3f2d5](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/1a3f2d588a8e26fea96920ea4152397a479440e6).

### Architecture Overview

The SAP JCo Connector architecture comprises key components facilitating communication between Ballerina applications and SAP systems. These components include the SAP function modules, the SAP system itself, the Ballerina code that handles the RFC calls, and the IDoc servers responsible for sending and receiving IDocs. This architecture enables robust integration, supporting various enterprise operations.

## 2. Components

This section outlines the core components of the Ballerina SAP JCo Connector and their configurations, focusing on how they interact with the SAP system.

### 2.1 RFC Client

The RFC Client component allows the Ballerina application to communicate with an SAP system by invoking remote function modules and sending IDocs.

#### 2.1.1 Configurations

The RFC Client supports two types of configurations, `DestinationConfig` and `AdvancedConfig`, to establish a connection with the SAP system.

##### 2.1.1.1 DestinationConfig

The `DestinationConfig` type represents the configuration details needed to create a RFC connection.

```ballerina

> public type DestinationConfig record {]

Reviewer: Shall we include the relevant doc-comments too ?

Author: Added with [79c8a96](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/79c8a96dc8a6e16ce658e975cdae1ef3271ed12b).

Reviewer: Shall we add the doc comments like this ?

```ballerina
public type DestinationCoinfig record {|
    # The SAP host name (jco.client.ashost)
    string ashost;
    // other fields
|};
```

Author: Changed with [fbf51f1](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/fbf51f16e882bc2a682303a3488c9f4381f7f2c6).

    string ashost;
    string sysnr;
    string jcoClient;
    string user;
    string passwd;
    string lang = "EN";
    string group = "PUBLIC";
};
```

**Example Configuration:**

```ballerina
configurable jco:DestinationConfig config = ?;
jco:Client sapClient = check new (config);
```

Config.toml file:

```toml
[config]
ashost = "host"
sysnr = "00"
jcoClient = "000"
user = "user"
passwd = "password"
lang = "en"
group = "group"
```

##### 2.1.1.2 AdvancedConfig

If the user wants to specify more detailed configurations beyond `DestinationConfig`, they can use `AdvancedConfig`. The `AdvancedConfig` type is a map that holds any SAP JCo configurations accepted by the SAP destination provider. 

**Example Configuration:**

```ballerina
configurable jco:AdvancedConfig config = ?;
jco:Client sapClient = check new (config);
```

Config.toml file:

```toml
[config]
"jco.destination.pool_capacity" = "3"
"jco.destination.peak_limit" = "10"
```

#### 2.1.2 Available Operations

##### 2.1.2.1 Execute a Function Module via RFC

An RFC function call is made using the following function;

```ballerina

> remote function execute(string functionName, record {|FieldType?...;|} importParams, typedesc<record {|FieldType?...;|}|xml|json?> exportParams = <>) returns exportParams|Error

Reviewer: ditto

Author: Added with [79c8a96](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/79c8a96dc8a6e16ce658e975cdae1ef3271ed12b).

### Architecture Overview

The SAP JCo Connector architecture comprises key components facilitating communication between Ballerina applications and SAP systems. These components include the SAP function modules, the SAP system itself, the Ballerina code that handles the RFC calls, and the IDoc servers responsible for sending and receiving IDocs. This architecture enables robust integration, supporting various enterprise operations.

## 2. Components

This section outlines the core components of the Ballerina SAP JCo Connector and their configurations, focusing on how they interact with the SAP system.

### 2.1 RFC Client

The RFC Client component allows the Ballerina application to communicate with an SAP system by invoking remote function modules and sending IDocs.

#### 2.1.1 Configurations

The RFC Client supports two types of configurations, `DestinationConfig` and `AdvancedConfig`, to establish a connection with the SAP system.

##### 2.1.1.1 DestinationConfig

The `DestinationConfig` type represents the configuration details needed to create a RFC connection.

```ballerina
public type DestinationConfig record {]

>     # The SAP hostname (jco.client.ashost).
>     string ashost;
>     # The SAP system number (jco.client.sysnr).
>     string sysnr;
>     # The SAP client (jco.client.client).
>     string jcoClient;
>     # The SAP user name (jco.client.user).
>     string user;
>     # The SAP password (jco.client.passwd).
>     string passwd;
>     # The SAP language (jco.client.lang).
>     string lang = "EN";
>     # The SAP group (jco.client.group).
>     string group = "PUBLIC";

Reviewer: ```suggestion
    # The SAP hostname (jco.client.ashost)
    string ashost;
    # The SAP system number (jco.client.sysnr)
    string sysnr;
    # The SAP client (jco.client.client)
    string jcoClient;
    # The SAP user name (jco.client.user)
    string user;
    # The SAP password (jco.client.passwd)
    string passwd;
    # The SAP language (jco.client.lang)
    string lang = "EN";
    # The SAP group (jco.client.group)
    string group = "PUBLIC";
```

};
```

**Example Configuration:**

```ballerina
configurable jco:DestinationConfig config = ?;
jco:Client sapClient = check new (config);
```

Config.toml file:

```toml
[config]
ashost = "host"
sysnr = "00"
jcoClient = "000"
user = "user"
passwd = "password"
lang = "en"
group = "group"
```

##### 2.1.1.2 AdvancedConfig

> If the user wants to specify more detailed configurations beyond `DestinationConfig`, they can use `AdvancedConfig`. The `AdvancedConfig` type is a map that holds any SAP JCo configurations accepted by the SAP destination provider. 

Reviewer: Can you include the `AdvancedConfig` record definition as well ?

Author: Included with [5fb5481](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/5fb5481cd4cd3236393eb4704d2c54fd71908db8).

**Example Configuration:**

```ballerina
configurable jco:AdvancedConfig config = ?;
jco:Client sapClient = check new (config);
```

Config.toml file:

```toml
[config]
"jco.destination.pool_capacity" = "3"
"jco.destination.peak_limit" = "10"
```

#### 2.1.2 Available Operations

##### 2.1.2.1 Execute a Function Module via RFC

An RFC function call is made using the following function;

```ballerina
# Executes the RFC function.
#

> # + functionName - The name of the function to be executed.
> # + importParams - The input parameters for the function.
> # + exportParams - The output parameters for the function.
> # + return - An error if the execution fails.

Reviewer: Check for other places as well.
```suggestion
# + functionName - The name of the function to be executed
# + importParams - The input parameters for the function
# + exportParams - The output parameters for the function
# + return - An error if the execution fails
```

Users can invoke it as shown below:

```ballerina
public function main() returns error? {
    ImportParams importParams = {
        importParam1: "Hello",
        importParam2: 123,
        importParam3: 123.45,
        importParam4: 123.456
    };

    ExportParams? result = check jcoClient->execute("TEST_FUNCTION", importParams);
    if result is jco:ExportParams {
        io:println("Result: ", result);
    } else {
        io:println("Error: Function execution failed");
    }
}
```

##### 2.1.2.2 Send IDocs to an SAP System

The RFC Client also supports sending IDocs to an SAP system, allowing you to automate the exchange of structured data.

An IDoc can be sent using the following function:

```ballerina
# Send the iDoc.
#
# + iDoc - The XML string of the iDoc.
# + iDocType - The type of the iDoc.
# + return - An error if the execution fails.

> remote function sendIDoc(xml iDoc, IDocType iDocType = DEFAULT) returns Error

Reviewer: ```suggestion
remote function sendIDoc(xml iDoc, IDocType iDocType = jco:DEFAULT) returns Error
```

```

For the export parameters, the type descriptor of the user's return type is extracted. It should be a closed record of `FieldType` as the rest fields or a closed record with the exact output parameter names from the SAP system. This initial record provided by the user will be considered the export parameter list, and within this closed record, the user can include nested records for `SAP structures` and record arrays for `SAP tables`.

Users can invoke it as shown below:

```ballerina
public function main() returns error? {
    ImportParams importParams = {
        importParam1: "Hello",
        importParam2: 123,
        importParam3: 123.45,
        importParam4: 123.456
    };

    ExportParams? result = check jcoClient->execute("TEST_FUNCTION", importParams);
    if result is jco:ExportParams {
        io:println("Result: ", result);
    } else {
        io:println("Error: Function execution failed");
    }
}
```

##### 2.1.2.2 Send IDocs to an SAP System

The RFC Client also supports sending IDocs to an SAP system, allowing you to automate the exchange of structured data.

An IDoc can be sent using the following function:

```ballerina

> # Send the iDoc.
> #
> # + iDoc - The XML string of the iDoc.
> # + iDocType - The type of the iDoc.
> # + return - An error if the execution fails.

Reviewer: ```suggestion
# Send the iDoc.
#
# + iDoc - The XML string of the iDoc
# + iDocType - The type of the iDoc
# + return - An error if the execution fails
```

Author: Fixed with [fde31e6](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/fde31e60abca0e591cb6c1685ee6591702a5ccbb).

| string                                | char[]         | STRING, SSTRING       |
| time:Date                             | Date           | DATE                  |
| time:TimeOfDay                        | Time           | TIME                  |
| int                                   | short          | INT1                  |
| int                                   | int            | INT2                  |
| int                                   | long           | INT4                  |
| float                                 | float          | FLOAT                 |
| decimal                               | double         | DEC, QUAN, CURR, PREC |
| byte                                  | byte           | INT1                  |
| byte[]                                | byte[]         | RAW, LRAW, BYTE       |
| decimal                               | BigDecimal     | DEC, QUAN, CURR       |
| int                                   | BigInteger     | NUM                   |
| record (with/without rest fields)     | JCoStructure   | STRUCTURE             |
| record[] (with/without rest fields)   | JCoTable       | TABLE                 |

> ### 2.2 Listener

Reviewer: We have not included the functions defined in the `jco:Listener`. Shall we add those too ?

Author: Changed with [cf4c263](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/cf4c263ee0ab616f3fac2aa71fcabf05d0712c17).

#### 2.2.1 Configurations

The Listener supports two types of configurations, `ServerConfig` and `AdvancedConfig`, to facilitate communication with the SAP system by receiving IDocs and handling events.

##### 2.2.1.1 ServerConfig

The `ServerConfig` type represents the configuration details needed to create an IDoc connection.

```ballerina
public type ServerConfig record {
    # The gateway host (jco.server.gwhost)
    string gwhost;
    # The gateway service (jco.server.gwserv)
    string gwserv;
    # The program ID (jco.server.progid)
    string progid;
};
```

**Example Configuration:**

```ballerina
configurable jco:ServerConfig configs = ?;
listener jco:Listener idocListener = new (configs);
```

Config.toml file:

```toml
[configs]
gwhost = "gwhost"
gwserv = "sapgw00"
progid = "progid"
```

> ##### 2.2.1.2 AdvancedConfig

Reviewer: If this is the same thing we defined previously, no need to redefine.

Author: Yes, but in this case, 'That' is used to accept a different set of values from the Client. The initial thought was that we need to demonstrate this using an example, even for the developer, since it plays a critical role in certain SAP ECC scenarios. In some versions of SAP ECC, this can be used to create a Destination (which happens only if specific configurations are provided), where a client and a listener are initiated in the background.

Reviewer: In that case this should have been something like this IMO;

#### 1. Using `AdvanceConfig` as field in the `ServerConfig`

```ballerina
public type ServerConfig record {
    # The gateway host (jco.server.gwhost)
    string gwhost;
    # The gateway service (jco.server.gwserv)
    string gwserv;
    # The program ID (jco.server.progid)
    string progid;
    # doc comment.....
    AdvancedConfig advanceConfig?;
};
```

#### 2. By opening the `ServerConfig` for `string` fields

Anyway `ServerConfig` is an open-record, this is always allowed.

Both the above options assumes that `gwhost`, `gwserv`, and `progid` are required fields in both regular and advance cases. 

Any reason to not to do that ? (again this is more like an issue with the design, but better to clarify it at this point)

Author: We followed this approach as it is much more user-friendly for both new and existing users.

Existing users already have a `.jcoDestination` file that contains a set of configurations like the following:

```
#Standard destination
jco.client.client=MANDANT
jco.client.ashost=APPLICATION_SERVER_HOST
jco.client.sysnr=INSTANZNUMMER
jco.client.r3name=SYSTEMID
jco.client.saprouter=SAP_ROUTER_STRING
jco.client.lang=DE
jco.server.mshost=
jco.client.user=USERNAME
jco.client.passwd=PASSWORD
jco.server.gwhost=GWHOST.COM
jco.server.gwserv=SAPGW00
jco.server.progid=PROGRAMID
jco.server.connection_count=1
jco.server.repository_destination=POOL2
```

This can easily be represented as a Ballerina map, which is why we introduced the `AdvancedConfig` record.

On the other hand, new users (using the later version of the SDK) will only need to obtain the following information from their SAP system administrator to create a listener. which reduces the cognitive load on the new users.

```
gwhost = GWHOST.COM
gwserv = SAPGW00
progid = PROGRAMID
```

However, the method suggested in '2. Opening the ServerConfig for string fields' is also a good option, and we will discuss it in future improvement planning meetings.

# Specification: Ballerina SAP JCo Connector

_Authors_: @RDPerera\
_Reviewers_: @daneshk @aashikam @shafreenAnfar\
_Created_: 2024/08/28\

> _Updated_: 2024/09/02\

Reviewer: ```suggestion
_Updated_: 2024/09/06\
```

The conforming implementation of the specification is released and included in the distribution. Any deviation from the specification is considered a bug.

## Contents

1. [Overview](#1-overview)
2. [Components](#2-components)
   - 2.1 [RFC Client](#21-rfc-client)
     - 2.1.1 [Configurations](#211-configurations)
       - 2.1.1.1 [DestinationConfig](#2111-destinationconfig)
       - 2.1.1.2 [AdvancedConfig](#2112-advancedconfig)
     - 2.1.2 [Available Operations](#212-available-operations)
       - 2.1.2.1 [Execute a Function Module via RFC](#2121-execute-a-function-module-via-rfc)
       - 2.1.2.2 [Send IDocs to an SAP System](#2122-send-idocs-to-an-sap-system)
     - 2.1.3 [Data Types](#213-data-type-mapping-between-ballerina-and-sap-jco)
   - 2.2 [Listener](#22-listener)
     - 2.2.1 [Configurations](#221-configurations)
       - 2.2.1.1 [ServerConfig](#2211-serverconfig)
       - 2.2.1.2 [AdvancedConfig](#2212-advancedconfig)
     - 2.2.2 [Available Operations](#222-available-operations)
       - 2.2.2.1 [Receive IDocs](#2221-receive-idocs)

>        - 2.2.2.2 [Error Handling in Listener](#2222-error-handling)

Reviewer 4: IINM  we recently agreed to use sentence case for titles as a convention across all new library content. So I believe this should be applied to the spec as well. @ThisaruGuruge @daneshk wdyt?

Reviewer 3: Yes, let's use sentence case

Author: Changed with [0553b90](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/0553b901a67bf7a546060ea221052c9ae955cb00)

### Architecture Overview

The SAP JCo Connector architecture comprises key components facilitating communication between Ballerina applications and SAP systems. These components include the SAP function modules, the SAP system itself, the Ballerina code that handles the RFC calls, and the IDoc servers responsible for sending and receiving IDocs. This architecture enables robust integration, supporting various enterprise operations.

## 2. Components

This section outlines the core components of the Ballerina SAP JCo Connector and their configurations, focusing on how they interact with the SAP system.

### 2.1 RFC Client

The RFC Client component allows the Ballerina application to communicate with an SAP system by invoking remote function modules and sending IDocs.

#### 2.1.1 Configurations

The RFC Client supports two types of configurations, `DestinationConfig` and `AdvancedConfig`, to establish a connection with the SAP system.

> ##### 2.1.1.1 DestinationConfig

Reviewer 4: Wouldn't it be better to have the title as `Destination configurations` (instead of having just the record typename) for better readability?
```suggestion
##### 2.1.1.1 DestinationConfig
```

The `DestinationConfig` type represents the configuration details needed to create a RFC connection.

```ballerina
public type DestinationConfig record {]
    # The SAP hostname (jco.client.ashost)
    string ashost;
    # The SAP system number (jco.client.sysnr)
    string sysnr;
    # The SAP client (jco.client.client)
    string jcoClient;
    # The SAP user name (jco.client.user)
    string user;
    # The SAP password (jco.client.passwd)
    string passwd;
    # The SAP language (jco.client.lang)
    string lang = "EN";
    # The SAP group (jco.client.group)
    string group = "PUBLIC";
};
```

**Example Configuration:**

```ballerina
configurable jco:DestinationConfig config = ?;
jco:Client sapClient = check new (config);
```

Config.toml file:

```toml
[config]
ashost = "host"
sysnr = "00"
jcoClient = "000"
user = "user"
passwd = "password"
lang = "en"
group = "group"
```

> ##### 2.1.1.2 AdvancedConfig

Reviewer 4: ditto

Author: Yes, sure fixed with [0553b90](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/0553b901a67bf7a546060ea221052c9ae955cb00).

```

The input parameters require functionName and importParams. The functionName refers to the Remote Function Module name and importParams accepts a FieldType value, which can be;

```ballerina
public type FieldType string|int|float|decimal|time:Date|time:TimeOfDay|byte[]|record {|FieldType?...;|}|record {|FieldType?...;|}[];
```

For the export parameters, the type descriptor of the user's return type is extracted. It should be a closed record of `FieldType` as the rest fields or a closed record with the exact output parameter names from the SAP system. This initial record provided by the user will be considered the export parameter list, and within this closed record, the user can include nested records for `SAP structures` and record arrays for `SAP tables`.

Users can invoke it as shown below:

```ballerina
public function main() returns error? {
    ImportParams importParams = {
        importParam1: "Hello",
        importParam2: 123,
        importParam3: 123.45,
        importParam4: 123.456
    };

    ExportParams? result = check jcoClient->execute("TEST_FUNCTION", importParams);

>     if result is jco:ExportParams {

Reviewer 2: Do we have an in-built record type for `ExportParams` in jco?

Author: No.

Author: We are returning the data according to the user-given `ExportParams` type.

```
remote function execute(string functionName, record {|FieldType?...;|} importParams, typedesc<record {|FieldType?...;|}|xml|json?> exportParams = <>) returns exportParams|Error
```

| int                                   | int            | INT2                  |
| int                                   | long           | INT4                  |
| float                                 | float          | FLOAT                 |
| decimal                               | double         | DEC, QUAN, CURR, PREC |
| byte                                  | byte           | INT1                  |
| byte[]                                | byte[]         | RAW, LRAW, BYTE       |
| decimal                               | BigDecimal     | DEC, QUAN, CURR       |
| int                                   | BigInteger     | NUM                   |
| record (with/without rest fields)     | JCoStructure   | STRUCTURE             |
| record[] (with/without rest fields)   | JCoTable       | TABLE                 |

### 2.2 Listener

#### 2.2.1 Configurations

The Listener supports two types of configurations, `ServerConfig` and `AdvancedConfig`, to facilitate communication with the SAP system by receiving IDocs and handling events.

##### 2.2.1.1 ServerConfig

The `ServerConfig` type represents the configuration details needed to create an IDoc connection.

```ballerina

> public type ServerConfig record {

Reviewer: Any reason to mark the `ServerConfig` as an open record ? Usual practice is to use closed records.

Author: Yes. Fixed with https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/57.

```ballerina
public type AdvancedConfig map<string>;
```

**Example Configuration:**

```ballerina
configurable jco:AdvancedConfig configs = ?;
listener jco:Listener idocListener = new (configs);
```

Config.toml file:

```toml
[configs]
"jco.server.gwhost" = "gwhost"
"jco.server.gwserv" = "sapgw00"
"jco.server.repository_destination" = "SAMPLE_DESTINATION"
"jco.client.ashost" = "host"
"jco.client.sysnr" = "00"
"jco.client.client" = "000"
"jco.client.user" = "user"
"jco.client.passwd" = "password"
```
Here's how the updated section could look, incorporating the provided code:

#### 2.2.2 Functions

This section describes how to attach, detach, start, and stop the listener, along with initializing it for listening to iDoc messages.

- To attach a service to the iDoc listener, the `attach` function can be used.

```ballerina
# Attaches a `Service` to the iDoc listener.
# ```
# check idocListener.attach(service);
# ```
#
# + `s` - The service instance to be attached.
# + `name` - The name of the service (optional).
# + `return` - An error if the attachment fails or else `()`.

> public isolated function attach(Service s, string[]|string? name = ()) returns error? = @java:Method {

Reviewer: This question is not directly regarding the spec. The usual practice is to define a module level error type and use it in anywhere there can be an error return. Any reason to deviate from that here ?

Author: Yes, fixed with https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/57.

# ```
# check idocListener.init(config);
# ```
#
# + `configurations` - The required configuration for initializing the listener.
# + `serverName` - Name of the server (optional, default is a UUID).
# + `return` - An error if initialization fails or else `()`.
public isolated function init(ServerConfig|AdvancedConfig serverConfig, string serverName = uuid:createType4AsString()) returns Error? {
    return externInit(self, serverConfig, serverName);
}
```

#### 2.2.3 Available Operations

##### 2.2.3.1 Receive IDocs

The Listener can be used to receive IDocs from the SAP system, which can then be processed within your Ballerina application.

```ballerina
listener jco:Listener iDocListener = new (configurations);

service jco:Service on iDocListener {
    remote function onReceive(xml iDoc) returns error? {
        check io:fileWriteXml("resources/received_idoc.xml", iDoc);
        io:println("IDoc received and saved.");
    }
}
```

When an IDoc is received, it will be in XML format, and the user can easily map it to a record using Ballerina's XML-to-Record conversion features.

##### 2.2.3.2 Error Handling

>  Listener
> 
> The Listener also supports error handling, allowing you to capture and manage any issues that occur during IDoc reception.

Reviewer 4: Is this word added intentionally? looks redundant to me
```suggestion
The Listener also supports error handling, allowing you to capture and manage any issues that occur during IDoc reception.
```

Author: Changed with [0553b90](https://github.com/ballerina-platform/module-ballerinax-sap.jco/pull/56/commits/0553b901a67bf7a546060ea221052c9ae955cb00).
